File: backend/app/models/game.py
from pydantic import BaseModel
from typing import Optional, List
from enum import Enum
from .character import Character, CharacterType

class ActionType(str, Enum):
    ATTACK = "attack"
    DEFEND = "defend"
    SKILL = "skill"

class GameState(str, Enum):
    CHARACTER_SELECT = "character_select"
    PLAYER_TURN = "player_turn"
    AI_TURN = "ai_turn"
    FINISHED = "finished"

class Player(BaseModel):
    id: str
    name: str
    character: Optional[Character] = None

class AIOpponent(BaseModel):
    character: Character
    difficulty: str = "normal"

class GameSession(BaseModel):
    session_id: str
    player: Player
    ai_opponent: Optional[AIOpponent] = None
    state: GameState = GameState.CHARACTER_SELECT
    current_turn: int = 1
    is_player_turn: bool = True
    turn_history: List[str] = []
    winner: Optional[str] = None

    def is_ready(self) -> bool:
        return (self.player.character is not None and
                self.ai_opponent is not None)

class GameAction(BaseModel):
    action_type: ActionType
    damage_dealt: int = 0
    damage_taken: int = 0
    defense_value: int = 0

class SelectCharacterRequest(BaseModel):
    character_type: CharacterType

class GameActionRequest(BaseModel):
    action_type: ActionType

class StartGameRequest(BaseModel):
    player_name: str

File: frontend/src/types/game.ts
export enum CharacterType {
  WARRIOR = "warrior",
  TANKER = "tanker",
  MAGE = "mage"
}

export enum ActionType {
  ATTACK = "attack",
  DEFEND = "defend",
  SKILL = "skill"
}

export enum GameState {
  CHARACTER_SELECT = "character_select",
  PLAYER_TURN = "player_turn",
  AI_TURN = "ai_turn",
  FINISHED = "finished"
}

export interface Character {
  name: string;
  character_type: CharacterType;
  max_hp: number;
  current_hp: number;
  attack_range: [number, number];
  defense_range: [number, number];
  skill_name: string;
  turns_since_last_skill: number;
  skill_cooldown: number;
}

export interface Player {
  id: string;
  name: string;
  character?: Character;
}

export interface AIOpponent {
  character: Character;
  difficulty: string;
}

export interface GameSession {
  session_id: string;
  player: Player;
  ai_opponent?: AIOpponent;
  state: GameState;
  current_turn: number;
  is_player_turn: boolean;
  turn_history: string[];
  winner?: string;
}

export interface GameMessage {
  type: string;
  [key: string]: any;
}